trellis-triplestore/src/main/java/org/trellisldp/triplestore/TriplestoreResourceService.java

```diff
@@ -231,20 +241,45 @@ private Boolean storeAndNotify(final IRI identifier, final Dataset dataset, fina
         return unmodifiableList(versions);
     }
 
+    private static final Predicate<BlankNodeOrIRI> isUserGraph = Trellis.PreferUserManaged::equals;
+    private static final Predicate<BlankNodeOrIRI> isServerGraph = Trellis.PreferServerManaged::equals;
 
-    private void emitEvents(final IRI identifier, final Literal time, final Dataset dataset) {
+    private void emitEvents(final IRI identifier, final Session session, final OperationType opType,
+            final Literal time, final Dataset dataset) {
 
         // Get the base URL
-        final Optional<String> baseUrl = dataset.getGraph().stream(identifier, DC.isPartOf, null)
-            .map(Triple::getObject).map(term -> ((IRI) term).getIRIString()).findAny();
+        final Optional<String> baseUrl = session.getProperty(TRELLIS_SESSION_BASE_URL);
+        final IRI inbox = dataset.getGraph(Trellis.PreferUserManaged)
+            .flatMap(graph -> graph.stream(null, LDP.inbox, null).map(Triple::getObject)
+                    .filter(term -> term instanceof IRI).map(term -> (IRI) term).findFirst())
+            .orElse(null);
+        final List<IRI> targetTypes = dataset.stream()
+            .filter(quad -> quad.getGraphName().filter(isUserGraph.or(isServerGraph)).isPresent())
+            .filter(quad -> quad.getPredicate().equals(RDF.type))
+            .flatMap(quad -> quad.getObject() instanceof IRI ? Stream.of((IRI) quad.getObject()) : empty())
+            .distinct().collect(toList());
 
         eventService.ifPresent(svc -> {
-            svc.emit(new SimpleEvent(getUrl(identifier, baseUrl), dataset));
+            svc.emit(new SimpleEvent(getUrl(identifier, baseUrl),
+                        asList(session.getAgent()), asList(PROV.Activity, getIriFromOperationType(opType)),
+                        targetTypes, inbox));
             getContainer(identifier).ifPresent(parent ->
-                    emitEventsForAdjacentResources(svc, parent, time, baseUrl, dataset));
+                    emitEventsForAdjacentResources(svc, parent, session, opType, time));
         });
     }
 
+    private IRI getIriFromOperationType(final OperationType opType) {
```

Comment:
Should we put this in `OperationType` (as a `public` constant or as a `static` method in the `enum`)?

Reply:
Yes, that makes sense to me.